### ISSUE TITLE
IllegalArgumentException

### ISSUE THREAD

**Issue by leima888** [DBCG/cql_engine#18]
In the cql library I use (col.elm.xml) effectiveDateTime is checked against a certain range to decide whether an Observation should be included in the report while the effectiveDateTime can be null in an Observation, in which case the Observation should theoritically be excluded from the report. However, when I ran the report with certain Observations having null effectiveDateTimes the following runtime exception is thrown and no report is generated:

java.lang.IllegalArgumentException: Errors occurred attempting to invoke the accessor function for property effectiveDateTimeType of type Observation
	org.opencds.cqf.cql.data.fhir.BaseFhirDataProvider.resolveProperty(BaseFhirDataProvider.java:238)

I did some digging and found that the exception is caused by an exception thrown in the Observation class (org.hl7.fhir.dstu3.model):

    public DateTimeType getEffectiveDateTimeType() throws FHIRException { 
      if (!(this.effective instanceof DateTimeType))
        throw new FHIRException("Type mismatch: the type DateTimeType was expected, but "+this.effective.getClass().getName()+" was encountered");
      return (DateTimeType) this.effective;
    }

The above method throws a FHIRException instead of returning null when the target is null. This is not a loner. There're quite a number of accessors in the hapi fhri model classes following this pattern.

**Comment by c-schuler** [comment]
I am unable to reproduce the error you are experiencing with null effectiveDateTime values... 

Which version of the engine are you using?

We actually have a test in the engine that uses the col.elm.xml library to evaluate a measure:
https://github.com/DBCG/cql_engine/blob/master/cql-engine-fhir/src/test/java/org/opencds/cqf/cql/data/fhir/TestFhirMeasureEvaluator.java#L42

The test runs the following CQL expression, which checks an effectiveDateTime against a range:
https://github.com/DBCG/cql_engine/blob/master/cql-engine-fhir/src/test/resources/org/opencds/cqf/cql/data/fhir/library-col.cql#L108

**Comment by leima888** [comment]
I'm using the engine of the version of master-SNAPSHOT and hapi fhir 2.3 (the dependencies down below). Have you ever tested against hapi fhir 2.3 (the official release of FHIR)?
		<dependency>
			<groupId>com.github.DBCG</groupId>
			<artifactId>cql_engine</artifactId>
			<version>master-SNAPSHOT</version>
		</dependency>
		<dependency>
			<groupId>ca.uhn.hapi.fhir</groupId>
			<artifactId>hapi-fhir-base</artifactId>
			<version>2.3</version>
		</dependency>

		<dependency>
		  <groupId>ca.uhn.hapi.fhir</groupId>
		  <artifactId>hapi-fhir-structures-dstu3</artifactId>
		  <version>2.3</version>
		</dependency>
		<dependency>
		  <groupId>ca.uhn.hapi.fhir</groupId>
		  <artifactId>hapi-fhir-validation-resources-dstu3</artifactId>
		  <version>2.3</version>
		</dependency>

**Comment by c-schuler** [comment]
Yeah, so you are using Jitpack or some equivalent... 
I would highly recommend using our SNAPSHOT releases with OSSRH (https://oss.sonatype.org/content/repositories/snapshots/org/opencds/cqf/). Those will provide a much more stable build for your project. 

You'll need the following in your pom.xml:
```
<repository>
        <id>oss-sonatype</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
</repository>

<dependency>
        <groupId>org.opencds.cqf</groupId>
        <artifactId>cql-engine</artifactId>
        <version>1.2.14-SNAPSHOT</version>
</dependency>
        
<dependency>
	<groupId>org.opencds.cqf</groupId>
	<artifactId>cql-engine-fhir</artifactId>
	<version>1.2.14-SNAPSHOT</version>
</dependency>
```
The SNAPSHOT versions are tracked here:
https://github.com/DBCG/cql_engine/wiki/1.2.0-SNAPSHOT-Release#snapshots

The current version of the Engine uses HAPI 2.4, which is the release for FHIR 3.0.0 (DSTU3). The 1.2.3-SNAPSHOT of the engine used HAPI 2.3 though.

**Comment by leima888** [comment]
I have tried the 1.2.14-SNAPSHOT release with HAPI 2.4 and get the same exception when effectiveDateTime is null. I don't seem to see any changes in the pieces where the exception is caused.

The following method in the Observation class (org.hl7.fhir.dstu3.model) throws a FHIRException when effective is null:
public DateTimeType getEffectiveDateTimeType() throws FHIRException { 
  if (!(this.effective instanceof DateTimeType))
    throw new FHIRException("Type mismatch: the type DateTimeType was expected, but "+this.effective.getClass().getName()+" was encountered");
  return (DateTimeType) this.effective;
}

The following method in  org.opencds.cqf.cql.data.fhir.BaseFhirDataProvider.resolveProperty(BaseFhirDataProvider.java) catches the exception and throw a java.lang.IllegalArgumentException: Errors occurred attempting to invoke the accessor function for property effectiveDateTimeType of type Observation
org.opencds.cqf.cql.data.fhir.BaseFhirDataProvider.resolveProperty(BaseFhirDataProvider.java)

protected Object resolveProperty(Object target, String path) {
        //........
        try {
            String accessorMethodName = String.format("%s%s%s", "get", path.substring(0, 1).toUpperCase(), path.substring(1));
            String elementAccessorMethodName = String.format("%sElement", accessorMethodName);
            Method accessor = null;
            try {
                accessor = clazz.getMethod(elementAccessorMethodName);
            }
            catch (NoSuchMethodException e) {
                accessor = clazz.getMethod(accessorMethodName);
            }

            Object result = accessor.invoke(target);
            result = mapPrimitive(result, target);
            return result;
        } catch (NoSuchMethodException e) {
            if (pathIsChoice(path)) {
                return resolveChoiceProperty(target, path);
            }
            else {
                throw new IllegalArgumentException(String.format("Could not determine accessor function for property %s of type %s", path, clazz.getSimpleName()));
            }
        } catch (InvocationTargetException e) {
            throw new IllegalArgumentException(String.format("Errors occurred attempting to invoke the accessor function for property %s of type %s", path, clazz.getSimpleName()));
        } catch (IllegalAccessException e) {
            throw new IllegalArgumentException(String.format("Could not invoke the accessor function for property %s of type %s", path, clazz.getSimpleName()));
        }
    }

**Comment by c-schuler** [comment]
Resolved: https://github.com/DBCG/cql_engine/commit/1971e6a821152c8c1109657c5ab1c54d77d488e2

Get the 1.2.15-SNAPSHOT version of the engine for the fix.

I was able to reproduce the issue you're experiencing. However, I got a different error message than what you're getting (Errors occurred attempting to invoke the accessor function for property effectiveDateTimeType of type Observation). I get the following error: Could not determine accessor function for property effectiveDateTimeType of type Observation. That error stems from resolving choice property types.

The issue was not with the getEffectiveDateTimeType() method throwing a FHIRException. That method is never accessed during execution because effectiveDateTime is a choice type. The engine resolves choice types by accessing the "root" (getEffective()) and comparing to the type (getEffective() instanceof DateTime). There was a bug in the choice type resolution logic, which was trying to verify the type DateTime is in the hapi dstu3 model package. However, DateTime is not in that package... DateTimeType is. I added some logic to append "Type" to the type name if the type cannot be found. That should fix your issue. 

Let me know if that did the trick =)
